From 4c354c53a6475820c481b31f8fb44194b69aefc6 Mon Sep 17 00:00:00 2001
From: Thibault Guittet <88336850+rh-tguittet@users.noreply.github.com>
Date: Mon, 18 Dec 2023 12:07:32 +0100
Subject: [PATCH] Check string length correctly to prevent memory
 over-consumption (RHEL-19527)

pk_strlen() cannot return larger values than the sentinel value (1024)
in this case and make the return value check uniform.
---
 src/pk-engine.c      | 4 ++--
 src/pk-transaction.c | 4 ++--
 2 files changed, 4 insertions(+), 4 deletions(-)

diff --git a/src/pk-engine.c b/src/pk-engine.c
index c910c1f18..f5e7c78d8 100644
--- a/src/pk-engine.c
+++ b/src/pk-engine.c
@@ -686,7 +686,7 @@ pk_engine_set_proxy (PkEngine *engine,
 
 	/* check length of http */
 	len = pk_strlen (proxy_http, 1024);
-	if (len == 1024) {
+	if (len >= 1024) {
 		error = g_error_new_literal (PK_ENGINE_ERROR,
 					     PK_ENGINE_ERROR_CANNOT_SET_PROXY,
 					     "http proxy was too long");
@@ -696,7 +696,7 @@ pk_engine_set_proxy (PkEngine *engine,
 
 	/* check length of ftp */
 	len = pk_strlen (proxy_ftp, 1024);
-	if (len == 1024) {
+	if (len >= 1024) {
 		error = g_error_new_literal (PK_ENGINE_ERROR,
 					     PK_ENGINE_ERROR_CANNOT_SET_PROXY,
 					     "ftp proxy was too long");
diff --git a/src/pk-transaction.c b/src/pk-transaction.c
index 5c24462c7..f67d48345 100644
--- a/src/pk-transaction.c
+++ b/src/pk-transaction.c
@@ -2484,7 +2484,7 @@ pk_transaction_strvalidate (const gchar *text, GError **error)
 				     "Invalid input passed to daemon: zero length string");
 		return FALSE;
 	}
-	if (length > 1024) {
+	if (length >= 1024) {
 		g_set_error (error, PK_TRANSACTION_ERROR, PK_TRANSACTION_ERROR_INPUT_INVALID,
 			     "Invalid input passed to daemon: input too long: %u", length);
 		return FALSE;
@@ -2536,7 +2536,7 @@ pk_transaction_search_check_item (const gchar *values, GError **error)
 				     "Invalid search containing '?'");
 		return FALSE;
 	}
-	if (size == 1024) {
+	if (size >= 1024) {
 		g_set_error_literal (error,
 				     PK_TRANSACTION_ERROR,
 				     PK_TRANSACTION_ERROR_SEARCH_INVALID,
